porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go

```diff
@@ -90,349 +90,161 @@ func (r *PackageVariantSetReconciler) Reconcile(ctx context.Context, req ctrl.Re
-	upstreamPackageName string) *pkgvarapi.Downstream {
-	downstreamPackageName := target.Package.Name
-	if downstreamPackageName == "" {
-		downstreamPackageName = upstreamPackageName
-	}
-	return &pkgvarapi.Downstream{
-		Repo:    target.Package.Repo,
-		Package: downstreamPackageName,
-	}
-}
+	for i, target := range pvs.Spec.Targets {
+		if len(target.Repositories) > 0 {
+			for _, rt := range target.Repositories {
+				pns := []string{upstreamPackageName}
+				if len(rt.PackageNames) > 0 {
+					pns = rt.PackageNames
+				}
 
-func (r *PackageVariantSetReconciler) repositorySet(
-	target *api.Target,
-	upstreamPackageName string,
-	repoList *configapi.RepositoryList) ([]*pkgvarapi.Downstream, error) {
-	var result []*pkgvarapi.Downstream
-	for _, repo := range repoList.Items {
-		repoAsRNode, err := r.convertObjectToRNode(&repo)
-		if err != nil {
-			return nil, fmt.Errorf("error converting repo to RNode: %v", err)
-		}
-		downstreamPackageName, err := r.getDownstreamPackageName(target.PackageName, upstreamPackageName, repoAsRNode)
-		if err != nil {
-			return nil, err
+				for _, pn := range pns {
+					result = append(result, pvContext{
+						template:       target.Template,
+						repoDefault:    rt.Name,
+						packageDefault: pn,
+					})
+				}
+			}
+			continue
 		}
-		result = append(result, &pkgvarapi.Downstream{
-			Repo:    repo.Name,
-			Package: downstreamPackageName,
-		})
 
-	}
-	return result, nil
-}
+		objSel := target.ObjectSelector
+		if target.RepositorySelector != nil {
+			// a label selector against a set of repositories
+			// equivlanet to object selector with apiVersion/kind pre-set
 
-func (r *PackageVariantSetReconciler) objectSet(target *api.Target,
-	upstreamPackageName string,
-	selectedObjects map[resid.ResId]*yaml.RNode) ([]*pkgvarapi.Downstream, error) {
-	var result []*pkgvarapi.Downstream
-	for _, obj := range selectedObjects {
-		downstreamPackageName, err := r.getDownstreamPackageName(target.PackageName,
-			upstreamPackageName, obj)
-		if err != nil {
-			return nil, err
-		}
-		repo, err := r.fetchValue(target.Objects.RepoName, obj)
-		if err != nil {
-			return nil, err
-		}
-		if repo == "" {
-			return nil, fmt.Errorf("error evaluating repo name: received empty string")
+			objSel = &api.ObjectSelector{
+				LabelSelector: *target.RepositorySelector,
+				APIVersion:    configapi.TypeRepository.APIVersion(),
+				Kind:          configapi.TypeRepository.Kind,
+			}
 		}
-		result = append(result, &pkgvarapi.Downstream{
-			Package: downstreamPackageName,
-			Repo:    repo,
-		})
-	}
-	return result, nil
-}
-
-func (r *PackageVariantSetReconciler) getSelectedObjects(ctx context.Context, selectors []api.Selector) (map[resid.ResId]*yaml.RNode, error) {
-	selectedObjects := make(map[resid.ResId]*yaml.RNode) // this is a map to prevent duplicates
-
-	for _, selector := range selectors {
+		// a selector against a set of arbitrary objects
```

Comment:
Is the rest of this supposed to be under a `if objSel != nil`? 

That, or maybe you're missing a `continue` under the `target.RepositorySelector` condition, as I see you have a `continue` under `target.Repositories`?

Reply:
No, it's correct. It's a little hard to tell because of all the change, but the logic of the interior of the loop is this:

 - if it's "repositories", handle it and continue
 - if it's repository selector, do the same as "object selector" but set GVK to be a porch Repository

So there are really only two sets of logic - repositories and object selectors, with repository selector just setting the default object info

Reply:
Ah got it! Thank you for the explanation.